src/ssl.c

```diff
@@ -14741,34 +14775,61 @@ int wolfSSL_Cleanup(void)
 #ifndef NO_SESSION_CACHE
     #ifdef ENABLE_SESSION_CACHE_ROW_LOCK
     for (i = 0; i < SESSION_ROWS; ++i) {
-        if (wc_FreeMutex(&SessionCache[i].row_mutex) != 0)
-            ret = BAD_MUTEX_E;
+        if ((SessionCache[i].mutex_valid == 1) &&
+            (wc_FreeMutex(&SessionCache[i].row_mutex) != 0)) {
+            if (ret == WOLFSSL_SUCCESS)
+                ret = BAD_MUTEX_E;
+        }
+        else {
+            SessionCache[i].mutex_valid = 0;
+        }
     }
     #else
-    if (wc_FreeMutex(&session_mutex) != 0)
-        ret = BAD_MUTEX_E;
+    if ((session_mutex_valid == 1) && (wc_FreeMutex(&session_mutex) != 0)) {
+        if (ret == WOLFSSL_SUCCESS)
+            ret = BAD_MUTEX_E;
+    }
+    else {
+        session_mutex_valid = 0;
+    }
     #endif
     #ifndef NO_CLIENT_CACHE
-        if (wc_FreeMutex(&clisession_mutex) != 0)
-            ret = BAD_MUTEX_E;
+        if ((clisession_mutex_valid == 1) &&
+            (wc_FreeMutex(&clisession_mutex) != 0)) {
+            if (ret == WOLFSSL_SUCCESS)
+                ret = BAD_MUTEX_E;
+        }
+        else {
+            clisession_mutex_valid = 0;
+        }
     #endif
 #endif /* !NO_SESSION_CACHE */
 
-    if (wc_FreeMutex(&count_mutex) != 0)
-        ret = BAD_MUTEX_E;
+    if ((count_mutex_valid == 1) && (wc_FreeMutex(&count_mutex) != 0)) {
+        if (ret == WOLFSSL_SUCCESS)
+            ret = BAD_MUTEX_E;
+    }
+    else {
+        count_mutex_valid = 0;
+    }
 
 #ifdef OPENSSL_EXTRA
     wolfSSL_RAND_Cleanup();
 #endif
 
     if (wolfCrypt_Cleanup() != 0) {
         WOLFSSL_MSG("Error with wolfCrypt_Cleanup call");
-        ret = WC_CLEANUP_E;
+        if (ret == WOLFSSL_SUCCESS)
+            ret = WC_CLEANUP_E;
     }
 
 #ifdef HAVE_GLOBAL_RNG
-    if (wc_FreeMutex(&globalRNGMutex) != 0) {
-        ret = BAD_MUTEX_E;
+    if ((globalRNGMutex_valid == 1) && (wc_FreeMutex(&globalRNGMutex) != 0)) {
+        if (ret == WOLFSSL_SUCCESS)
+            ret = BAD_MUTEX_E;
+    }
+    else {
+        globalRNGMutex_valid = 0;
```

Comment:
We should always set the "is valid" flags to 0 if the free fails. Same for the other flags.